test/gurka/test_recost.cc

```diff
@@ -0,0 +1,309 @@
+
+  // lets also do a bunch of costings
+  Api api;
+  auto json = actor.route(R"({"costing":"auto","locations":[)" + locations + R"(],"recostings":[
+      {"costing":"auto","name":"same"},
+      {"costing":"auto","name":"avoid_highways","use_highways":0.1},
+      {"costing":"bicycle","name":"slower"},
+      {"costing":"pedestrian","name":"slower_still"},
+      {"costing":"pedestrian","name":"slowest","walking_speed":2}]})",
+                          {}, &api);
+
+  auto greater_equal = [](const valhalla::TripLeg::PathCost& lesser,
+                          const valhalla::TripLeg::PathCost& greater, bool cost = true,
+                          bool transition = true) {
+    if (cost)
+      EXPECT_GE(greater.elapsed_cost().cost(), lesser.elapsed_cost().cost());
+    EXPECT_GE(greater.elapsed_cost().seconds(), lesser.elapsed_cost().seconds());
+    if (transition) {
+      if (cost)
+        EXPECT_GE(greater.transition_cost().cost(), lesser.transition_cost().cost());
+      EXPECT_GE(greater.transition_cost().seconds(), lesser.transition_cost().seconds());
+    }
+  };
+
+  // check we have the right amount of costing information at all places
+  for (const auto& n : api.trip().routes(0).legs(0).node()) {
+    EXPECT_EQ(n.recosts_size(), 5);
+    greater_equal(n.cost(), n.recosts(0));
+    greater_equal(n.recosts(0), n.recosts(1));
+    // this should be strickly bigger because of avoid highways
+    if (n.cost().elapsed_cost().seconds() > 0)
+      EXPECT_GT(n.recosts(1).elapsed_cost().cost(), n.recosts(0).elapsed_cost().cost());
+    greater_equal(n.recosts(1), n.recosts(2), true, false);
+    // bike cost uses different costing units so we only do the seconds
+    greater_equal(n.recosts(2), n.recosts(3), false, false);
+    greater_equal(n.recosts(3), n.recosts(4));
+  }
+
+  // TODO: verify json has the right info
+
+  // TODO: do it again for osrm format
+  json = actor.route(R"({"costing":"auto","format":"osrm", "locations":[)" + locations +
+                         R"(],"recostings":[
+      {"costing":"auto","name":"same"},
+      {"costing":"auto","name":"avoid_highways","use_highways":0.1},
+      {"costing":"bicycle","name":"slower"},
+      {"costing":"pedestrian","name":"slower_still"},
+      {"costing":"pedestrian","name":"slowest","walking_speed":2}]})",
+                     {}, &api);
+
+  api.Clear();
+  json = actor.route(R"({"costing":"auto","locations":[)" + locations + R"(],"recostings":[
+      {"costing":"truck","name":"nope"},
+      {"costing":"auto","name":"same"}]})",
+                     {}, &api);
+
+  // the first recosting should always be blank
+  for (const auto& n : api.trip().routes(0).legs(0).node()) {
+    EXPECT_EQ(n.recosts_size(), 2);
+    EXPECT_FALSE(n.recosts(0).has_elapsed_cost());
+    EXPECT_FALSE(n.recosts(0).has_transition_cost());
+    EXPECT_TRUE(n.recosts(1).has_elapsed_cost());
+    EXPECT_TRUE(n.recosts(1).has_transition_cost());
+  }
+
+  // TODO: verify json has the right info
```

Comment:
we should also check the json before we merge this. shouldnt be too much code to write but we need to check all the properties show up in all the right places